### PR TITLE
Add UVM specializations for iamax and scal

### PR DESCRIPTION
**Pull request by vqd8a**
Addressing #758. This is needed by ADELUS and GEMMA.

### CODE REVIEW

**Comment by ndellingwood** [comment]
@vqd8a please post results of spot-check and spot-check-tpls from kokkos-dev-2 when finished, thanks

**Comment by vqd8a** [comment]
@ndellingwood I do not have access to kokkos-dev-2. Can you help me with spot-checks? Thanks.

**Comment by ndellingwood** [comment]
@vqd8a sure! I'll launch them now

**Comment by ndellingwood** [comment]
Test results:

kokkos-dev-2 spot-check:
```bash
Running on machine: kokkos-dev-2
KokkosKernels Repository Status:  4368ef99f684c2f466a64bcc65357398e77e668c Add UVM specializations for iamax and scal

Kokkos Repository Status:  a00371d63a296048a043a31f2eaf440e249c4b36 Merge pull request #3130 from masterleinad/fix_nvcc_link_warning


Going to test compilers:  gcc/7.3.0 gcc/8.3.0 gcc/9.1 gcc/4.8.4 intel/17.0.1 intel/18.0.5 intel/19.0.5 clang/8.0 clang/9.0.0 cuda/10.1 cuda/9.2
...
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=629 run_time=164
clang-8.0-Pthread_Serial-release build_time=215 run_time=107
clang-9.0.0-Pthread-release build_time=121 run_time=48
clang-9.0.0-Serial-release build_time=125 run_time=54
cuda-10.1-Cuda_OpenMP-release build_time=809 run_time=162
cuda-9.2-Cuda_Serial-release build_time=788 run_time=191
gcc-4.8.4-OpenMP-release build_time=111 run_time=57
gcc-7.3.0-OpenMP-release build_time=143 run_time=58
gcc-7.3.0-Pthread-release build_time=122 run_time=52
gcc-8.3.0-Serial-release build_time=138 run_time=58
gcc-9.1-OpenMP-release build_time=178 run_time=58
gcc-9.1-Serial-release build_time=157 run_time=58
intel-17.0.1-Serial-release build_time=251 run_time=56
intel-18.0.5-OpenMP-release build_time=351 run_time=52
intel-19.0.5-Pthread-release build_time=436 run_time=49
```

kokkos-dev-2 spot-check-tpls:
```bash
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=633 run_time=168
clang-8.0-Pthread_Serial-release build_time=208 run_time=108
clang-9.0.0-Pthread-release build_time=133 run_time=51
clang-9.0.0-Serial-release build_time=128 run_time=57
cuda-10.1-Cuda_OpenMP-release build_time=848 run_time=167
gcc-4.8.4-OpenMP-release build_time=121 run_time=59
gcc-7.3.0-OpenMP-release build_time=143 run_time=59
gcc-7.3.0-Pthread-release build_time=120 run_time=52
gcc-8.3.0-Serial-release build_time=145 run_time=59
gcc-9.1-OpenMP-release build_time=187 run_time=59
gcc-9.1-Serial-release build_time=159 run_time=58
intel-17.0.1-Serial-release build_time=248 run_time=55
intel-18.0.5-OpenMP-release build_time=363 run_time=51
intel-19.0.5-Pthread-release build_time=447 run_time=47
```

**Comment by ndellingwood** [comment]
@vqd8a spot-check results posted, let me know when the PR is ready based on resolution of https://github.com/kokkos/kokkos-kernels/issues/758#issuecomment-653226329

**Comment by vqd8a** [comment]
Thanks @ndellingwood. 
Based on Christian's comment in https://github.com/kokkos/kokkos-kernels/issues/758#issuecomment-653240379, should this PR get merged ?@srajama1

**Comment by srajama1** [comment]
Sorry, I have even more questions. If some one enables UVMSpace with ETI, do they get only portion of the kernels and not all the kernels ?  Our typical style is when we support a memory space for ETI all our kernels are ETId for it. Also I would like this not to be default. Folks have to explicitly say they want it.

**Comment by vqd8a** [comment]
If UVMSpace with ETI is enabled (which is default in CMake logic when Kokkos is CUDA-enabled), I believe we get all the kernel pre-instantiated. However for cuBLAS TPL, we currently have CudaUVMSpace specializations only for TRSM, TRMM, GEMM, TRTRI. Other BLAS kernels are using their fall-back implementations with CudaUVMSpace. This PR adds CudaUVMSpace specializations for cuBLAS's IAMAX and SCAL.

> Also I would like this not to be default. Folks have to explicitly say they want it.

@srajama1 Did you mean users have to say what kernels need ETId?

I am wondering if we can use something like:
```
#ifdef KOKKOSKERNELS_INST_MEMSPACE_CUDAUVMSPACE
KOKKOSBLAS1_IAMAX_TPL_SPEC_AVAIL_CUBLAS( unsigned long, Kokkos::complex<double>,Kokkos::LayoutLeft, Kokkos::CudaUVMSpace)
KOKKOSBLAS1_SCAL_TPL_SPEC_AVAIL_CUBLAS( Kokkos::complex<double>, Kokkos::LayoutLeft, Kokkos::CudaUVMSpace)
#endif
``` 
for the CudaUVMSpace specializations.

File: src/impl/tpls/KokkosBlas1_iamax_tpl_spec_avail.hpp [hunk 1]
```diff
@@ -107,18 +107,26 @@ Kokkos::View<const SCALAR*, LAYOUT, Kokkos::Device<ExecSpace, MEMSPACE>, \
 #if defined (KOKKOSKERNELS_INST_DOUBLE)
 KOKKOSBLAS1_IAMAX_TPL_SPEC_AVAIL_CUBLAS( unsigned long, double,                 Kokkos::LayoutLeft, Kokkos::CudaSpace)
 KOKKOSBLAS1_IAMAX_TPL_SPEC_AVAIL_CUBLAS( unsigned int, double,                  Kokkos::LayoutLeft, Kokkos::CudaSpace)
+KOKKOSBLAS1_IAMAX_TPL_SPEC_AVAIL_CUBLAS( unsigned long, double,                 Kokkos::LayoutLeft, Kokkos::CudaUVMSpace)
```
Comment (jjwilke) [inline src/impl/tpls/KokkosBlas1_iamax_tpl_spec_avail.hpp:110]: I think we will want to guard this with a `KOKKOS_UVM_ENABLED` ifdef?

Comment (vqd8a) [reply src/impl/tpls/KokkosBlas1_iamax_tpl_spec_avail.hpp:110]: @jjwilke Yes. Do we already have ```KOKKOS_UVM_ENABLED``` in Kokkos or Kokkos Kernels? or will you add it to CMake?

Comment (jjwilke) [reply src/impl/tpls/KokkosBlas1_iamax_tpl_spec_avail.hpp:110]: Yeah, it comes from Kokkos itself. It's `KOKKOS_ENABLE_CUDA_UVM`, sorry, had the wrong name. As long as we include `Kokkos_Core.hpp` we will have that macro.

Comment (vqd8a) [reply src/impl/tpls/KokkosBlas1_iamax_tpl_spec_avail.hpp:110]: @jjwilke One question: what is the difference between ```KOKKOSKERNELS_INST_MEMSPACE_CUDAUVMSPACE``` and ```KOKKOS_UVM_ENABLED```?

Comment (vqd8a) [reply src/impl/tpls/KokkosBlas1_iamax_tpl_spec_avail.hpp:110]: Thanks, @jjwilke

**Comment by vqd8a** [comment]
As @jjwilke suggested, I added if guards with ```KOKKOS_ENABLE_CUDA_UVM```.
@ndellingwood could you please help me with spot-checks? Thanks.
@srajama1 @ndellingwood I think this PR is ready to merge when it passes spot-checks.

**Comment by ndellingwood** [comment]
Tests running now

**Comment by vqd8a** [comment]
Thanks, @ndellingwood

**Comment by ndellingwood** [comment]
Tested SHAs

```bash
KokkosKernels Repository Status:  4c1245996176edab577eb8cbb4d7002a6d657125 Add if guard KOKKOS_ENABLE_CUDA_UVM

Kokkos Repository Status:  4fa4b86a88fe44251c6bf466541e7e8c64cf0d66 Merge pull request #3164 from masterleinad/suppress_openmptarget_warning
```

**kokkos-dev-2 spot-check**
```bash
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=619 run_time=162
clang-8.0-Pthread_Serial-release build_time=205 run_time=101
clang-9.0.0-Pthread-release build_time=124 run_time=49
clang-9.0.0-Serial-release build_time=125 run_time=52
cuda-10.1-Cuda_OpenMP-release build_time=793 run_time=158
cuda-9.2-Cuda_Serial-release build_time=755 run_time=183
gcc-4.8.4-OpenMP-release build_time=114 run_time=57
gcc-7.3.0-OpenMP-release build_time=142 run_time=57
gcc-7.3.0-Pthread-release build_time=116 run_time=48
gcc-8.3.0-Serial-release build_time=139 run_time=56
gcc-9.1-OpenMP-release build_time=177 run_time=57
gcc-9.1-Serial-release build_time=160 run_time=55
intel-17.0.1-Serial-release build_time=263 run_time=54
intel-18.0.5-OpenMP-release build_time=382 run_time=52
intel-19.0.5-Pthread-release build_time=431 run_time=44
```

**kokkos-dev-2 spot-check-tpls**
```bash
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=617 run_time=163
clang-8.0-Pthread_Serial-release build_time=212 run_time=113
clang-9.0.0-Pthread-release build_time=121 run_time=49
clang-9.0.0-Serial-release build_time=136 run_time=54
cuda-10.1-Cuda_OpenMP-release build_time=811 run_time=165
gcc-4.8.4-OpenMP-release build_time=116 run_time=56
gcc-7.3.0-OpenMP-release build_time=140 run_time=58
gcc-7.3.0-Pthread-release build_time=120 run_time=50
gcc-8.3.0-Serial-release build_time=138 run_time=56
gcc-9.1-OpenMP-release build_time=176 run_time=58
gcc-9.1-Serial-release build_time=167 run_time=57
intel-17.0.1-Serial-release build_time=245 run_time=53
intel-18.0.5-OpenMP-release build_time=384 run_time=50
intel-19.0.5-Pthread-release build_time=437 run_time=45
```

**Comment by ndellingwood** [comment]
@vqd8a changes look good based on comments prior to your update.
@jjwilke can you look over the changes Vinh added and confirm your comments are addressed?

**Comment by ndellingwood** [comment]
I'll use the "squash and merge" option to squash the two commits to one and merge, thanks @vqd8a @jjwilke !

**Comment by vqd8a** [comment]
Thanks, @ndellingwood  and @jjwilke !

**Pull request merged**